https://redmine.buzzhives.com/issues/19975

lenganngoh opened the pull request.

sg-jsonjuliane commented on TripKitAndroidUI/src/main/java/com/skedgo/tripkit/ui/tripresult/UpdateTripForRealtime.kt:
```diff
@@ -15,70 +16,85 @@ import java.util.concurrent.atomic.AtomicReference
+            .withLatestFrom(
+                getTripGroup,
+                BiFunction<Long, TripGroup, TripGroup> { _, tripGroup -> tripGroup })
+            .filter { it.displayTrip!!.hasQuickBooking().not() }
```
### Code Review: Null Safety

The code currently relies on non-null assertions (`!!`) for certain properties, such as `it.displayTrip` and `it.displayTrip!!.updateURL`. To enhance the code's robustness and safety, consider the following improvements:

#### Use Safe Calls (?.)

Replace non-null assertions with safe calls to handle potential null values more gracefully. For example, instead of:

```kotlin
it.displayTrip!!.updateURL

sg-jsonjuliane commented on TripKitAndroidUI/src/main/java/com/skedgo/tripkit/ui/tripresult/UpdateTripForRealtime.kt:
```diff
@@ -15,70 +16,85 @@ import java.util.concurrent.atomic.AtomicReference
-        }
-  }
+    fun start(getTripGroup: Observable<TripGroup>) {
+        Observable.interval(0, 10, TimeUnit.SECONDS, Schedulers.computation())
```
### Code Review: Handling Magic Numbers

The code uses magic numbers, such as 0, 10, and `TimeUnit.SECONDS`, without clear explanations. To improve code readability and maintainability, consider the following suggestions:

####  Replace Magic Numbers with Constants

Define named constants for magic numbers to make the code more self-explanatory. For example, instead of using `0` and `10`, create constants like `INITIAL_DELAY_SECONDS` and `INTERVAL_SECONDS` with meaningful names.

#### Example:

```kotlin
// Define named constants for magic numbers
private val INITIAL_DELAY_SECONDS = 0
private val INTERVAL_SECONDS = 10

Observable.interval(INITIAL_DELAY_SECONDS, INTERVAL_SECONDS, TimeUnit.SECONDS, Schedulers.computation())

sg-jsonjuliane commented on TripKitAndroidUI/src/main/java/com/skedgo/tripkit/ui/tripresult/UpdateTripForRealtime.kt:
```diff
@@ -15,70 +16,85 @@ import java.util.concurrent.atomic.AtomicReference
+            .subscribe({
+                tripGroupRepository.updateTrip(
+                    it.second.uuid(),
+                    it.second.displayTrip!!.uuid(),
```
Please refer to: **Code Review: Null Safety**

sg-jsonjuliane commented on TripKitAndroidUI/src/main/java/com/skedgo/tripkit/ui/tripresult/UpdateTripForRealtime.kt:
```diff
@@ -15,70 +16,85 @@ import java.util.concurrent.atomic.AtomicReference
-        }
-  }
+    fun startForBooking(getTripGroup: Observable<TripGroup>) {
+        Observable.interval(0, 10, TimeUnit.SECONDS, Schedulers.computation())
```
Please refer to: **Code Review: Handling Magic Numbers**

sg-jsonjuliane commented on TripKitAndroidUI/src/main/java/com/skedgo/tripkit/ui/tripresult/UpdateTripForRealtime.kt:
```diff
@@ -15,70 +16,85 @@ import java.util.concurrent.atomic.AtomicReference
+            .subscribe({
+                tripGroupRepository.updateTrip(
+                    it.second.uuid(),
+                    it.second.displayTrip!!.uuid(),
```
Please refer to: **Code Review: Null Safety**

sg-jsonjuliane commented on TripKitAndroidUI/src/main/java/com/skedgo/tripkit/ui/tripresult/UpdateTripForRealtime.kt:
```diff
@@ -15,70 +16,85 @@ import java.util.concurrent.atomic.AtomicReference
+    private fun startAsync(group: TripGroup): Observable<Pair<Trip, TripGroup>> {
+        return Observable.just(group)
+            .observeOn(Schedulers.io())
+            .filter { !group.displayTrip!!.hasQuickBooking() }
```
Please refer to: **Code Review: Null Safety**

sg-jsonjuliane reviewed (commented):
One area with significant potential for improvement in this code review is the addition of comprehensive code comments and documentation. Enhancing code comments and documentation can greatly benefit code clarity, maintainability, and collaboration

sg-jsonjuliane reviewed (changes requested):
One area with significant potential for improvement in this code review is the addition of comprehensive code comments and documentation. Enhancing code comments and documentation can greatly benefit code clarity, maintainability, and collaboration

lenganngoh commented:
Updated PR with review comments. Removed the logs as they are used only for troubleshooting. Also incorporated requested null checks and magic numbers defining.

cc @sg-jsonjuliane
